text/0000-message-ordering-guarantee.md

```diff
@@ -0,0 +1,171 @@
+
+Example:
+
+```C#
+void FixedUpdate()
+{
+    GameObject obj = GameObject.Instantiate(prefab, location, rotation);
+    NetworkBehavior networkBehavior = obj.GetComponent<NetworkBehavior>();
+    networkBehavior.Spawn();
+    // Requires server ownership at this point, ChangeOwnership() must not have been processed yet
+    networkBehavior.SendAfterSpawnClientRPC();
+    networkBehavior.ChangeOwnership(newOwnerClientId);
+    // Requires client ownership, ChangeOwnership() must have been processed before this
+    networkBehavior.SendAfterChangeOwnershipClientRpc();
+}
+```
+
+In this example, without strict ordering, the `ChangeOwnership()` call may happen first, followed by `Spawn()`, followed thereafter by the two RPCs. Such an ordering would result in errors from the `ChangeOwnership()` call (due to the object not yet existing) and the second RPC (due to the ownership not having been changed). Ensuring that all messages arrive in the same order they were sent also ensures that they are called in the order required to process these four operations successfully.
+
+Another example:
+
+```C#
+void FixedUpdate()
+{
+    SendFixedUpdateClientRpc();
+}
+
+void Update()
+{
+    SendUpdateClientRpc();
+}
+```
+
+In this example, the client will execute `SendFixedUpdateClientRpc()` during the `FixedUpdate` stage, and `SendUpdateClientRpc()` during the `Update` stage - this ensures behavior between the sender and the recipient is somewhat more predictable, as logic that occurs on one side during `FixedUpdate` will implicitly occur on the other side during the same stage unless requested otherwise. 
+
+Likewise, the same is true of internal commands:
+
+```C#
+GameObject m_obj;
+
+void Update()
+{
+    m_obj = GameObject.Instantiate(prefab, location, rotation);
+    NetworkBehavior networkBehavior = m_obj.GetComponent<NetworkBehavior>();
+    networkBehavior.Spawn();
+}
+
+void LateUpdate()
+{
+    ulong newOwnerClientId = GetNewOwnerClientId();
+    NetworkBehavior networkBehavior = m_obj.GetComponent<NetworkBehavior>();
+    networkBehavior.ChangeOwnership(newOwnerClientId);
+}
+```
+
+In this example, because `Spawn()` is called during `Update()` on the server, and `ChangeOwnership()` is called during `LateUpdate()` on the server, the same will be true on the client - the object will be spawned on the client during the `Update` stage, and its ownership changed during `LateUpdate`.
+
+# Reference-level explanation
```

Comment:
I think there would be a wider and more fundamental refactor which would be to remove `InternalMessageSender` and `InternalMessageHandler` classes entirely. We stripped down both classes, especially `InternalMessageSender` is currently very slim. If we could also make these two types a part of this refactor work, that'd be super nice as well. I believe we could kill IMS easily, IMH maybe not so easily but having a look at that IMH class and its methods would be nice, at least to give us an idea about what it'd take to kill it.
Having said that, we'll ultimately formalize our network frame that starts with "network input" and ends with "network output" — during a frame everything would be just queued to our network output buffer. We're not there yet but I see this step being a positive step towards that direction, I like it!

Reply:
I think it'll definitely be easy to remove IMS with these changes. I'll investigate removing IMH as well.